// @okta/okta-angular/esm2015/src/okta/okta.module.js
/*
 * Copyright (c) 2017-Present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */
import * as tslib_1 from "tslib";
import { NgModule, Inject, Optional } from '@angular/core';
import { Router } from '@angular/router';
import { Location } from '@angular/common';
import { OktaCallbackComponent } from './components/callback.component';
import { OktaAuthGuard } from './okta.guard';
import { OktaAuthStateService } from './services/auth-state.service';
import { OktaHasAnyGroupDirective } from './has-any-group.directive';
import { OKTA_CONFIG } from './models/okta.config';
import { OktaAuth, AuthSdkError, toRelativeUrl } from '@okta/okta-auth-js';
import packageInfo from './packageInfo';
let OktaAuthModule = class OktaAuthModule {
    constructor(config, location, router) {
        const { oktaAuth } = config;
        if (!oktaAuth._oktaUserAgent) {
            throw new AuthSdkError('_oktaUserAgent is not available on auth SDK instance. Please use okta-auth-js@^5.3.1 or higher.');
        }
        // Auth-js version compatibility runtime check
        const oktaAuthVersion = oktaAuth._oktaUserAgent.getVersion();
        const majorVersion = +oktaAuthVersion.split('.')[0];
        if (packageInfo.authJSMajorVersion !== majorVersion) {
            throw new AuthSdkError(`Passed in oktaAuth is not compatible with the SDK, okta-auth-js version ${packageInfo.authJSMajorVersion}.x is the current supported version.`);
        }
        // Add Okta UA
        oktaAuth._oktaUserAgent.addEnvironment(`${packageInfo.name}/${packageInfo.version}`);
        // Provide a default implementation of `restoreOriginalUri`
        if (!oktaAuth.options.restoreOriginalUri && router && location) {
            oktaAuth.options.restoreOriginalUri = (_, originalUri) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                const baseUrl = window.location.origin + location.prepareExternalUrl('');
                const routePath = toRelativeUrl(originalUri || '/', baseUrl);
                router.navigateByUrl(routePath);
            });
        }
        // Start services
        oktaAuth.start();
    }
};
OktaAuthModule = tslib_1.__decorate([
    NgModule({
        declarations: [
            OktaCallbackComponent,
            OktaHasAnyGroupDirective,
        ],
        exports: [
            OktaCallbackComponent,
            OktaHasAnyGroupDirective,
        ],
        providers: [
            OktaAuthGuard,
            OktaAuthStateService,
            {
                provide: OktaAuth,
                useFactory(config) {
                    return config.oktaAuth;
                },
                deps: [OKTA_CONFIG]
            },
        ]
    }),
    tslib_1.__param(0, Inject(OKTA_CONFIG)),
    tslib_1.__param(1, Optional()),
    tslib_1.__param(2, Optional()),
    tslib_1.__metadata("design:paramtypes", [Object, Location,
        Router])
], OktaAuthModule);
export { OktaAuthModule };
